stl/inc/xmemory

```diff
@@ -776,6 +776,118 @@ _NODISCARD constexpr allocation_result<typename allocator_traits<_Alloc>::pointe
+//
+// Because of the second property, we can only mark poison up to the final aligned address before the true `last`.
+// Otherwise, we'd poison the memory _after_ `last` as well.
+// For the first property, we can assume that everything before `first` in the shadow memory section is valid
+// (since otherwise we couldn't mark `first` valid), and so we just return back the first address in
+// `first`'s shadow memory section.
+//
+// ### Example
+//
+// ```cpp
+// struct alignas(8) cat {
+//     int meow; // bytes [0, 4)
+//     char buffer[16]; // bytes [4, 20)
+//     int purr; // bytes [20, 24)
+// };
+// ```
+//
+// First, `meow` and `purr` are just regular data members, not container buffers, so they _must_ be valid.
+// Then, assume we want to poison all of `buffer`.
+// This would mean that, in a perfect world, we want something like:
+//
+// ```
+//    |  meow |               buffer                | purr  |
+//   [ v v v v p p p p ][ p p p p p p p p ][ p p p p v v v v ]
+//           sm1                sm2                sm3
+// ```
+//
+// However, note that by the rules above, `sm3` is not a valid shadow memory section; we always need
+// the valid bytes to come before the poison bytes. Thus, the closest we can actually get to it is:
+//
+// ```
+//    |  meow |               buffer                | purr  |
+//   [ v v v v p p p p ][ p p p p p p p p ][ v v v v v v v v ]
+//           sm1                sm2                sm3
+// ```
+//
+// We call `aligned = _Get_asan_aligned_first_end(cat.buffer, cat.buffer + 16);`, and we get back
+//
+// ```cpp
+// aligned = {
+//     ._First = &cat.meow,
+//     ._End = cat.buffer + 12,
+// };
+// ```
+//
+// Then, we poison as much of buffer as we can via
+//
+// ```cpp
+// __sanitizer_annotate_contiguous_container(
+//     aligned._First,
+//     aligned._End,
+//     cat.buffer,
+//     aligned._Clamp_to_end(cat.buffer + 16));
+// ```
+//
+// We are allowed to assume that `&cat.meow` is valid, since otherwise `cat.buffer + [0, 4)` could not be valid.
+// We cannot poison up to `cat.buffer + 16`, since then `&purr` could not be valid.
+// Thus, this results in the shadow memory state from the second example.
+_NODISCARD inline _AsanAlignedPointers _Get_asan_aligned_first_end(
+    const void* const _First, const void* const _Last) noexcept {
+    static constexpr uintptr_t _Mask = ~(_Asan_granularity - 1);
+
+    return {
+        reinterpret_cast<const void*>(reinterpret_cast<uintptr_t>(_First) & _Mask),
+        reinterpret_cast<const void*>(reinterpret_cast<uintptr_t>(_Last) & _Mask),
+    };
+}
+
+template <class _Container, class = void>
+_INLINE_VAR constexpr size_t _Container_allocation_minimum_alignment = alignof(typename _Container::value_type);
+
+template <class _Container>
+_INLINE_VAR constexpr size_t _Container_allocation_minimum_alignment<_Container,
+    void_t<decltype(_Container::allocator_type::_Minimum_allocation_alignment)>> =

```

Comment:
Since custom allocators will leave false negatives, we likely want to document that behavior at https://learn.microsoft.com/en-us/cpp/sanitizers/error-container-overflow?view=msvc-170 and perhaps allow folks to opt-in to the over-poisoning by adding `_Minimum_allocation_alignment` to their allocators (maybe rename it to be obviously ASAN related - I believe it's otherwise unused?).

Reply:
microsoft/cpp-docs#4274